<h1 align = "center" style = "color :blue">Projet Final de formation de la bibliothèque Pandas (Le projets consiste a analyser des donnees fournis par une université)</h1>

In [1]:
import pandas as pd
import numpy as np

## Objectif

Réaliser une analyse complète de données universitaires en utilisant **Pandas uniquement** pour toutes les opérations de traitement et d'analyse.

Vous disposez de 5 fichiers CSV :

- `students.csv` : informations sur les étudiants
- `courses.csv` : informations sur les matières
- `grades.csv` : notes obtenues
- `attendance.csv` : assiduité par semestre
- `exams.csv` : informations sur les examens


## Partie 1 : Découverte des données

Pour chaque fichier :

1. Importer le CSV.
2. Afficher les premières et dernières lignes.
3. Afficher le nombre de lignes et de colonnes.
4. Afficher les noms des colonnes.
5. Examiner les types de données.
6. Produire un résumé statistique.
7. Identifier les valeurs manquantes.
8. Identifier les doublons.

### A) - Importation des fichier csv

In [2]:
Etudiant = pd.read_csv("data/students.csv") # importation des infos sur l'etudiant 

cours =pd.read_csv("data/courses.csv") # Les infos sur les matiere de chaque etudiant

notes_obtenues = pd.read_csv("data/grades.csv") # Information sur les notes des etudiants

assidute = pd.read_csv("data/attendance.csv") # Assidute de l'etudiant par semestre

examen = pd.read_csv("data/exams.csv") # Informations sur les examens

### B) - Affichage des premières et dernières lignes. des data

#### a) - Information des etudiants

In [3]:
Etudiant.head() # affiche les Premiere lignes 

,student_id,nom,prenom,age,sexe,filiere,niveau
0,ETU1000,Mensah,Koffi,20.0,M,Logistique & Transport,L3
1,ETU1001,Agbeko,Akossiwa,24.0,M,Informatique & Systèmes,L2
2,ETU1002,Doe,Afi,25.0,F,Logistique & Transport,L1
3,ETU1003,Amegan,Dodzi,NaN,F,Génie Électrique,L2
4,ETU1004,Amegan,Mawuli,18.0,M,IA & Big Data,L1


In [4]:
Etudiant.tail() # affiche les derniere lignes 

,student_id,nom,prenom,age,sexe,filiere,niveau
795,ETU1795,Adjei,Ama,23.0,F,Génie Électrique,L3
796,ETU1796,Assogba,Sena,26.0,M,Génie Électrique,L3
797,ETU1797,Amegan,Mawuli,19.0,M,Informatique & Systèmes,L3
798,ETU1798,Lawson,Eyram,19.0,F,Génie Civil,L3
799,ETU1799,Amegan,Mawuli,NaN,M,Génie Civil,L2


In [5]:
#  1 - Nombre de ligne et colonne dans le Dataframe Etudiant
Etudiant.shape

(800, 7)

In [6]:
# 2- Afficher les noms des colonnes.

Etudiant.columns

Index(['student_id', 'nom', 'prenom', 'age', 'sexe', 'filiere', 'niveau'], dtype='object')

In [7]:
# 3 ) - Examiner les types de données.
Etudiant.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   student_id  800 non-null    object 
 1   nom         800 non-null    object 
 2   prenom      800 non-null    object 
 3   age         788 non-null    float64
 4   sexe        792 non-null    object 
 5   filiere     800 non-null    object 
 6   niveau      800 non-null    object 
dtypes: float64(1), object(6)
memory usage: 43.9+ KB


#### On remaque la colones des ages a pout types float ==> (impossible donc a corriger)

In [8]:
Etudiant.describe() # 4 - )Resume statistques 

,age
count,788.000000
mean,21.446701
std,2.855424
min,17.000000
25%,19.000000
50%,21.000000
75%,24.000000
max,26.000000


In [9]:
Etudiant.isnull().sum() # la colonne age  a 12 notes manquantes

student_id     0
nom            0
prenom         0
age           12
sexe           8
filiere        0
niveau         0
dtype: int64

In [10]:
# Verifications des doucblons
Etudiant.duplicated().sum() #==> 0 doublons

np.int64(0)

#### b) - Information sur les matiere

In [11]:
cours.head() #  affiche les premieres lignes 

,course_id,matiere,credits,semestre
0,MAT101,Mathématiques,4,S1
1,PHY101,Physique,3,S1
2,PROG101,Programmation,5,S1
3,STAT101,Statistiques,4,S1
4,ANG101,Anglais,2,S1


In [12]:
cours.tail() #  affiche les dernieres lignes 

,course_id,matiere,credits,semestre
5,DATA101,Data Science,5,S2
6,ALG101,Algorithmique,4,S2
7,ELEC101,Électricité,3,S2
8,GEST101,Gestion,2,S2
9,PROJ101,Projet,4,S2


In [13]:
#  1 - Nombre de ligne et colonne dans le Dataframe contenat les cours
cours.shape  #on a au total 10 cours

(10, 4)

In [14]:
# 2- Afficher les noms des colonnes.

cours.columns

Index(['course_id', 'matiere', 'credits', 'semestre'], dtype='object')

In [15]:
# 3 ) - Examiner les types de données.
cours.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   course_id  10 non-null     object
 1   matiere    10 non-null     object
 2   credits    10 non-null     int64 
 3   semestre   10 non-null     object
dtypes: int64(1), object(3)
memory usage: 452.0+ bytes


In [16]:
cours.describe() # 4 - )Resume statistques 

,credits
count,10.000000
mean,3.600000
std,1.074968
min,2.000000
25%,3.000000
50%,4.000000
75%,4.000000
max,5.000000


In [17]:
cours.isnull().sum()   ## Pas de valeur manquante dans chacuns des colonnes

course_id    0
matiere      0
credits      0
semestre     0
dtype: int64

In [18]:
# Verifications des doucblons
cours.duplicated().sum() #  ==> 0 doublons

np.int64(0)

#### c) - Information sur les notes des etudiants

In [19]:
notes_obtenues.head()   # affiche les premieres lignes 

,student_id,course_id,note
0,ETU1000,MAT101,6.03
1,ETU1000,PHY101,6.54
2,ETU1000,PROG101,8.00
3,ETU1000,STAT101,10.48
4,ETU1000,ANG101,8.31


In [20]:
notes_obtenues.tail()   # affiche les dernieres lignes 

,student_id,course_id,note
7995,ETU1799,DATA101,11.12
7996,ETU1799,ALG101,14.30
7997,ETU1799,ELEC101,15.23
7998,ETU1799,GEST101,15.48
7999,ETU1799,PROJ101,16.37


In [21]:
notes_obtenues.shape ## 8000 note pour 3 colonne

(8000, 3)

In [22]:
notes_obtenues.columns

Index(['student_id', 'course_id', 'note'], dtype='object')

In [23]:
notes_obtenues.info()  # Les notes sont imcomplets il existe donc des notes manquantes dans la colone note (voir ->donnes de infos)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   student_id  8000 non-null   object 
 1   course_id   8000 non-null   object 
 2   note        7870 non-null   float64
dtypes: float64(1), object(2)
memory usage: 187.6+ KB


In [24]:
notes_obtenues.describe() #Resume statistique

,note
count,7870.000000
mean,11.902784
std,3.345841
min,0.000000
25%,9.620000
50%,11.860000
75%,14.220000
max,20.000000


In [25]:
        # Valeuer manquantes :
notes_obtenues.isnull().sum() # ==> La colonnnes de note a 130 notes manquantes

student_id      0
course_id       0
note          130
dtype: int64

In [26]:
# Verifications des doublons
notes_obtenues.duplicated().sum()  # 0  doublons 

np.int64(0)

#### d) - information sur l'assiduté    de l'etudiant par semestre

In [27]:
assidute.head()   # affiche les premieres lignes 

,student_id,semestre,taux_presence,absences,retards
0,ETU1000,S1,88.22,6,3
1,ETU1000,S2,79.11,10,6
2,ETU1001,S1,98.29,1,2
3,ETU1001,S2,92.43,4,4
4,ETU1002,S1,77.95,11,1


In [28]:
assidute.tail()   # affiche les dernieres lignes 

,student_id,semestre,taux_presence,absences,retards
1595,ETU1797,S2,79.05,10,1
1596,ETU1798,S1,92.19,4,2
1597,ETU1798,S2,72.55,14,2
1598,ETU1799,S1,90.08,5,1
1599,ETU1799,S2,72.99,14,4


In [29]:
assidute.shape #1600 ligne pour 5 colonnes

(1600, 5)

In [30]:
assidute.columns

Index(['student_id', 'semestre', 'taux_presence', 'absences', 'retards'], dtype='object')

In [31]:
assidute.info() # d'apres info ==>la colonnes taux_presence a des informations manquan

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1600 entries, 0 to 1599
Data columns (total 5 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   student_id     1600 non-null   object 
 1   semestre       1600 non-null   object 
 2   taux_presence  1585 non-null   float64
 3   absences       1600 non-null   int64  
 4   retards        1600 non-null   int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 62.6+ KB


In [32]:
assidute.describe() #Resumer statistique de l'assidute des etudiants

,taux_presence,absences,retards
count,1585.000000,1600.000000,1600.000000
mean,82.107754,8.941875,2.553125
std,11.364506,5.675485,1.603133
min,46.320000,0.000000,0.000000
25%,74.370000,5.000000,1.000000
50%,82.540000,9.000000,2.000000
75%,90.780000,13.000000,3.000000
max,100.000000,27.000000,10.000000


In [33]:
assidute.isnull().sum() ## ==> La colonne taux_precesenc a 15 informations manquantes

student_id        0
semestre          0
taux_presence    15
absences          0
retards           0
dtype: int64

In [34]:
# Verifications des doucblons
assidute.duplicated().sum() # ==> zero doublons

np.int64(0)

#### e) - Informations sur les examens

In [35]:
examen.head()   # affiche les premieres lignes 

,course_id,date_examen,session,salle,nombre_candidats
0,MAT101,2025-06-02,Matin,Salle D,44
1,PHY101,2025-06-05,Matin,Salle A,86
2,PROG101,2025-06-08,Matin,Salle C,61
3,STAT101,2025-06-11,Après-midi,Salle A,47
4,ANG101,2025-06-14,Matin,Salle C,79


In [36]:
examen.tail()   # affiche les dernieres lignes 

,course_id,date_examen,session,salle,nombre_candidats
5,DATA101,2025-06-17,Après-midi,Salle D,85
6,ALG101,2025-06-20,Matin,Salle C,83
7,ELEC101,2025-06-23,Après-midi,Salle C,81
8,GEST101,2025-06-26,Après-midi,Salle C,82
9,PROJ101,2025-06-29,Matin,Salle C,47


In [37]:
examen.shape # 10 lignes pour 5 colonnes

(10, 5)

In [38]:
examen.columns

Index(['course_id', 'date_examen', 'session', 'salle', 'nombre_candidats'], dtype='object')

In [39]:
examen.info() # => Aucun souci avec ce dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   course_id         10 non-null     object
 1   date_examen       10 non-null     object
 2   session           10 non-null     object
 3   salle             10 non-null     object
 4   nombre_candidats  10 non-null     int64 
dtypes: int64(1), object(4)
memory usage: 532.0+ bytes


In [40]:
examen.describe() # Resume statistiques

,nombre_candidats
count,10.000000
mean,69.500000
std,17.665094
min,44.000000
25%,50.500000
50%,80.000000
75%,82.750000
max,86.000000


In [41]:
# Valeur manquantes ==> AUCUNE
examen.isnull().sum()

course_id           0
date_examen         0
session             0
salle               0
nombre_candidats    0
dtype: int64

In [42]:
# Verifications des doublons
examen.duplicated().sum() #==> 0 doublons

np.int64(0)

## Partie 2 : Nettoyage

Nettoyer les données et expliquer les décisions prises.

Vous devez notamment :

- traiter les valeurs manquantes ;
- vérifier les doublons ;
- corriger les types de données si nécessaire ;
- vérifier les valeurs incohérentes ;
- uniformiser les données textuelles ;
- convertir les dates correctement.


#### Nettoyage du dataframe Etudiant 

In [43]:
Etudiant.info() #Verification des types de donnes 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 7 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   student_id  800 non-null    object 
 1   nom         800 non-null    object 
 2   prenom      800 non-null    object 
 3   age         788 non-null    float64
 4   sexe        792 non-null    object 
 5   filiere     800 non-null    object 
 6   niveau      800 non-null    object 
dtypes: float64(1), object(6)
memory usage: 43.9+ KB


In [44]:
Etudiant.isnull().sum()
# Dans la colonne age on 12 ages manquant, et dans sexe 12

student_id     0
nom            0
prenom         0
age           12
sexe           8
filiere        0
niveau         0
dtype: int64

In [45]:
Etudiant["age"] = Etudiant["age"].fillna("inconnue") # Remplace l'age manquant par inconnue
Etudiant["sexe"] = Etudiant["sexe"].fillna("inconnue") # Remplace le sexe manquant par inconnue

In [46]:
Etudiant.isnull().sum() # Erreures corriges

student_id    0
nom           0
prenom        0
age           0
sexe          0
filiere       0
niveau        0
dtype: int64

 2- ) -IL n'a pas de doublon car dans le dataframe Etudiant --> revoir partie 1

<p>3 -) -Correction des types de donnees</p>

Pas de problemme avec les types de donnees

#### Netoyage du dataframe  cours

In [47]:
cours.isnull().sum() 

course_id    0
matiere      0
credits      0
semestre     0
dtype: int64

In [48]:
# Pas d'erreur avec le dataframe cours

#### Netoyage du dataframe notes_obtenues

In [49]:
notes_obtenues.isnull().sum() # ==> 130 note manquantes

student_id      0
course_id       0
note          130
dtype: int64

In [50]:
# Correction des valeur manquantes
notes_obtenues["note"] =notes_obtenues["note"].fillna(0)
# Au debut j'avais decider de remplacer par le mot "inconnu" pout les notes manquantes , 
#mais la realite ma rattrapper dans la partie 2,Donc je suis revenu remplacer par 0 avant de continuer , 
#je pouvais bien utiliser les conditions sur inconnue mais bon c'est trop de taff

In [51]:
notes_obtenues.isnull().sum() 

student_id    0
course_id     0
note          0
dtype: int64

In [52]:
# Erreur corriger

#### Netoyage du dataframe assidute

In [53]:
assidute.isnull().sum()

student_id        0
semestre          0
taux_presence    15
absences          0
retards           0
dtype: int64

In [54]:
assidute["taux_presence"] = assidute["taux_presence"].fillna("inconnue")  

In [55]:
assidute.isnull().sum()

student_id       0
semestre         0
taux_presence    0
absences         0
retards          0
dtype: int64

In [56]:
#Erreur corriger

#### Netoyage du dataframe examen

In [57]:
examen.isnull().sum()

course_id           0
date_examen         0
session             0
salle               0
nombre_candidats    0
dtype: int64

In [58]:
# Pas de valeur manquant

In [59]:
examen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10 entries, 0 to 9
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   course_id         10 non-null     object
 1   date_examen       10 non-null     object
 2   session           10 non-null     object
 3   salle             10 non-null     object
 4   nombre_candidats  10 non-null     int64 
dtypes: int64(1), object(4)
memory usage: 532.0+ bytes


In [60]:
examen.head()

,course_id,date_examen,session,salle,nombre_candidats
0,MAT101,2025-06-02,Matin,Salle D,44
1,PHY101,2025-06-05,Matin,Salle A,86
2,PROG101,2025-06-08,Matin,Salle C,61
3,STAT101,2025-06-11,Après-midi,Salle A,47
4,ANG101,2025-06-14,Matin,Salle C,79


## Partie 3 : Analyse des étudiants

Répondre avec Pandas aux questions suivantes :

1. Combien d'étudiants sont enregistrés ?
2. Combien y a-t-il d'étudiants dans chaque filière ?
3. Quelle est la répartition hommes/femmes ?
4. Quelle est la moyenne d'âge ?
5. Quel est l'âge minimum et maximum ?
6. Combien d'étudiants sont en L1, L2 et L3 ?
7. Quelle filière contient le plus d'étudiants ?


####  1) -  Nombre d'etudiant enregistrer

In [61]:
nbreEtudiant = Etudiant.shape[0]
print("--------------------------------------------------------------")
print(f"Le nombre total des etudiants enregistrer :{nbreEtudiant } Etudiants")
print("--------------------------------------------------------------")

--------------------------------------------------------------
Le nombre total des etudiants enregistrer :800 Etudiants
--------------------------------------------------------------


#### 2) - Nombre d'etudiant dans chaque  filière 

In [62]:
Etudiant.head()


,student_id,nom,prenom,age,sexe,filiere,niveau
0,ETU1000,Mensah,Koffi,20.0,M,Logistique & Transport,L3
1,ETU1001,Agbeko,Akossiwa,24.0,M,Informatique & Systèmes,L2
2,ETU1002,Doe,Afi,25.0,F,Logistique & Transport,L1
3,ETU1003,Amegan,Dodzi,inconnue,F,Génie Électrique,L2
4,ETU1004,Amegan,Mawuli,18.0,M,IA & Big Data,L1


In [63]:
Etudiant["filiere"].value_counts()

filiere
Informatique & Systèmes    200
IA & Big Data              189
Génie Électrique           150
Génie Civil                131
Logistique & Transport     130
Name: count, dtype: int64

In [64]:
# Les infos ci-dessuis affiche les infos des etudiants inscrits dans chaque filiere

In [65]:
dic = {
    "filiere" :["Informatique & Systèmes","IA & Big Data","Génie Électrique","Génie Civil","Logistique & Transport"],
    "Nombre_Etudiant" :[200, 189, 150, 131,130],
}
df= pd.DataFrame(dic)
df

,filiere,Nombre_Etudiant
0,Informatique & Systèmes,200
1,IA & Big Data,189
2,Génie Électrique,150
3,Génie Civil,131
4,Logistique & Transport,130


#### 3) - Quelle est la répartition hommes/femmes ?

In [66]:
Etudiant["sexe"].value_counts()

sexe
M           465
F           327
inconnue      8
Name: count, dtype: int64

In [67]:
df = pd.DataFrame({
    "SEXE" : ["Masculin", "Feminin", "sexe_inconnue"],
    "NombreEtudiants" :[465, 327, 8]
})
df["repartition (%)"] = (df["NombreEtudiants"]/800 )*100
df

,SEXE,NombreEtudiants,repartition (%)
0,Masculin,465,58.125
1,Feminin,327,40.875
2,sexe_inconnue,8,1.000


#### 4) - Quelle est la moyenne d'âge ?

In [68]:
moyenne_ages = Etudiant.loc[Etudiant["age"] != "inconnue", "age"].mean()

print("--------------------------------------------------------------")
print(f"Moyenne des ages des etudiants :{ int(moyenne_ages)} ans")
print("--------------------------------------------------------------")


--------------------------------------------------------------
Moyenne des ages des etudiants :21 ans
--------------------------------------------------------------


#### 5) - Quel est l'âge minimum et maximum ?

In [69]:

Age_max = Etudiant.loc[Etudiant["age"] != "inconnue", "age"].max()
Age_min = Etudiant.loc[Etudiant["age"] != "inconnue", "age"].min()

print("--------------------------------------------------------------")
print(f"ages maximum  des etudiants :{int(Age_max)} ans")
print(f"ages minimum  des etudiants :{int(Age_min)} ans")
print("--------------------------------------------------------------")


--------------------------------------------------------------
ages maximum  des etudiants :26 ans
ages minimum  des etudiants :17 ans
--------------------------------------------------------------


#### 6-) - Combien d'étudiants sont en L1, L2 et L3 ?

In [70]:
Etudiant["niveau"].value_counts()

niveau
L1    336
L2    276
L3    188
Name: count, dtype: int64

In [71]:
df =pd.DataFrame({
    "niveau" : ["Licence_1", "Licence_2", "Licence_3"],
    "nombreEtudiant" :[336, 276, 188]
})

<h3 align = "center" style = "color : blue">Resume des nombres d'etudiant dans chaque niveau de licence </h3>

In [72]:
df

,niveau,nombreEtudiant
0,Licence_1,336
1,Licence_2,276
2,Licence_3,188


#### 7- ) - Quelle filière contient le plus d'étudiants ?

In [73]:
Etudiant["filiere"].value_counts()

filiere
Informatique & Systèmes    200
IA & Big Data              189
Génie Électrique           150
Génie Civil                131
Logistique & Transport     130
Name: count, dtype: int64

##### La filiere ***Informatique & Systèmes*** a plus d'etudiant : ***200 Etudiant inscrits*** dans cette  filiere

## Partie 4 : Analyse des notes

À partir de `grades.csv` et `courses.csv` :

1. Calculer la moyenne de chaque matière.
2. Trouver la matière ayant la meilleure moyenne.
3. Trouver la matière ayant la plus faible moyenne.
4. Calculer le nombre de notes disponibles par matière.
5. Calculer la moyenne des notes par semestre.
6. Identifier les étudiants ayant les meilleures notes.
7. Identifier les étudiants ayant les plus faibles résultats.
8. Calculer la moyenne générale de chaque étudiant.
9. Classer les étudiants selon leur moyenne générale.
10. Déterminer le nombre d'étudiants ayant une moyenne générale >= 10.
11. Déterminer le nombre d'étudiants ayant une moyenne générale < 10.

In [74]:
cours.head()

,course_id,matiere,credits,semestre
0,MAT101,Mathématiques,4,S1
1,PHY101,Physique,3,S1
2,PROG101,Programmation,5,S1
3,STAT101,Statistiques,4,S1
4,ANG101,Anglais,2,S1


In [75]:
notes_obtenues.head()

,student_id,course_id,note
0,ETU1000,MAT101,6.03
1,ETU1000,PHY101,6.54
2,ETU1000,PROG101,8.00
3,ETU1000,STAT101,10.48
4,ETU1000,ANG101,8.31


#### 1 - ) Calcule de  la moyenne de chaque matière.

In [76]:
#Pour ce faire d'abord je vais fusionner les deux dataframes notes_obtenues, et cours 
Data_notes = pd.merge(cours, notes_obtenues, on = "course_id", how = "inner")

In [77]:
Data_notes.head()

,course_id,matiere,credits,semestre,student_id,note
0,MAT101,Mathématiques,4,S1,ETU1000,6.03
1,MAT101,Mathématiques,4,S1,ETU1001,12.16
2,MAT101,Mathématiques,4,S1,ETU1002,5.80
3,MAT101,Mathématiques,4,S1,ETU1003,14.07
4,MAT101,Mathématiques,4,S1,ETU1004,7.36


In [78]:
Data_notes["matiere"].unique()  # Element de la colonne de matiere sans doublons

array(['Mathématiques', 'Physique', 'Programmation', 'Statistiques',
       'Anglais', 'Data Science', 'Algorithmique', 'Électricité',
       'Gestion', 'Projet'], dtype=object)

In [79]:
# ha moi qui voulais utiliser la methode ci-dessus pour le calcul de moyenne de chaque matiere , tchieer, jamais de la vie , boucle for ma preférée

Dico_moyenne ={}
for i in Data_notes["matiere"].unique():
    moy = Data_notes.loc[Data_notes["matiere"] == i, "note"].mean()
    moyenne = round(float(moy), 2)
    Dico_moyenne[i] = moyenne
    
Dico_moyenne

{'Mathématiques': 11.11,
 'Physique': 10.99,
 'Programmation': 11.69,
 'Statistiques': 11.44,
 'Anglais': 12.58,
 'Data Science': 11.98,
 'Algorithmique': 11.33,
 'Électricité': 11.0,
 'Gestion': 12.22,
 'Projet': 12.75}

In [80]:
df_moyenne = pd.DataFrame(Dico_moyenne , index =["moyenne"])

<h3 align = "center" style = "color : blue">Resume des moyennes pour chaque matiere</h3>

In [81]:
df_moyenne

,Mathématiques,Physique,Programmation,Statistiques,Anglais,Data Science,Algorithmique,Électricité,Gestion,Projet
moyenne,11.11,10.99,11.69,11.44,12.58,11.98,11.33,11.0,12.22,12.75


#### 2-) Trouver la matière ayant la meilleure moyenne.

d'apres le dataframe df_moyenne la meilleur 12,75 la matiere est :projet

`Autre methode (Essayons de trouver la meilleur moyenne a travers un programme 
Pour ce faire je vais partir a base du dictionnaire  Dico_moyenne  que j;ai creer dans la question 1`

In [82]:
moy_max = 0
for matiere, moy in Dico_moyenne.items() :
    if moy > moy_max:
        moy_max =moy
print(f"La meilleur moyenne est {matiere} : {moy_max}")

La meilleur moyenne est Projet : 12.75


##### Le programme confirme notre reponse :***moyenne est : 12.75** , ***matiere :	Projet***

#### 3-) Trouver la matière ayant la plus faible moyenne.

`|Pour ce faire je peut bien preciser la plus faible moyenne a partir du dataframe df_moyenne mais , bon je vais le determiner 
a partit d'un programme python en partant du dictionnaire Dico_moyenne`

In [83]:
moy_min = list(Dico_moyenne.values())[0] # initialisation de la  premiere moyenne  du dictionnaire
matiere_min = ""

for matiere , moyenne in Dico_moyenne.items():
    if moyenne < moy_min:
        moy_min = moyenne
        matiere_min = matiere
        
print(f"Moyenne , {matiere_min} :{moy_min}")

Moyenne , Physique :10.99


In [84]:
# Un peu de galere pour ce bloque de code bizarement j'en ai mis du temps avant que le programme  n'affiche ce que je voulais

##### En definitive la matiere ayant la plus faible moyenne es , ***Physique :10.99 de moyenne***

#### 4-) Calculer le nombre de notes disponibles par matière.

In [85]:
Data_notes["matiere"].value_counts() # Frequence des matiere

matiere
Mathématiques    800
Physique         800
Programmation    800
Statistiques     800
Anglais          800
Data Science     800
Algorithmique    800
Électricité      800
Gestion          800
Projet           800
Name: count, dtype: int64

`Il faut noter que j'ai eu a faire du netoyage sur les notes , car le fichier csv qui contenait les notes , 
il y'avait des notes manquante , et moi j'ai fait le netoyage en remplacant ces notes manquante par la valeur 0 a ces endroit la`

**En conclusion :** il y'a ***800 notes*** disponible pour chaques matiere

#### 5-) Calculer la moyenne des notes par semestre.

In [86]:
Data_notes.head() # On travaille avec dataframe cours

,course_id,matiere,credits,semestre,student_id,note
0,MAT101,Mathématiques,4,S1,ETU1000,6.03
1,MAT101,Mathématiques,4,S1,ETU1001,12.16
2,MAT101,Mathématiques,4,S1,ETU1002,5.80
3,MAT101,Mathématiques,4,S1,ETU1003,14.07
4,MAT101,Mathématiques,4,S1,ETU1004,7.36


In [87]:
Data_notes["semestre"].value_counts()

semestre
S1    4000
S2    4000
Name: count, dtype: int64

On au total on a ***4000 notes par par semestre***

In [88]:
Data = Data_notes.loc[:, ["semestre","note"]]
Data.head()    # J'ai essayer de simplifier le dataframe en nouveau dataframe Data qui regroupe 
#            les donnes qui me sont necessaire pour resoudre a la question

,semestre,note
0,S1,6.03
1,S1,12.16
2,S1,5.80
3,S1,14.07
4,S1,7.36


In [89]:
moyenne_S1 = round(Data.loc[Data["semestre"] == "S1", "note"].mean(), 2) #J'ai arrondi les moyenne a 2 valeur apres la virgule
moyenne_S2 = round(Data.loc[Data["semestre"] == "S2", "note"].mean(), 2)

print(f"moyenne S1 :{moyenne_S1}\nmoyenne S2 :{moyenne_S2}")

moyenne S1 :11.56
moyenne S2 :11.86


|Pour une meilleur exploration je vais traduire les infos dans un nouveau dataframe`

In [90]:
df = pd.DataFrame({"Semestre_1" :moyenne_S1, "Semestre_2" :moyenne_S2}, index =["moyenne"])
  #Resumer des donnees  dans un dataframe

### Resume de la moyenne des notes par semestre

In [91]:
df

,Semestre_1,Semestre_2
moyenne,11.56,11.86


#### 6-) Identifier les étudiants ayant les meilleures notes.

 Cette questions je la comprend pas trop mais bon ma reponse sera de trier et voir 

In [92]:
TrierData = Data_notes.sort_values("note",ascending = False)

<h4 align = "center" style = "color : blue">Resume des 20 premiers etudiant ayant obtenu les meilleures notes</h4>

In [93]:
TrierData.head(20)

,course_id,matiere,credits,semestre,student_id,note
7950,PROJ101,Projet,4,S2,ETU1750,20.0
7869,PROJ101,Projet,4,S2,ETU1669,20.0
7883,PROJ101,Projet,4,S2,ETU1683,20.0
6553,GEST101,Gestion,2,S2,ETU1153,20.0
6522,GEST101,Gestion,2,S2,ETU1122,20.0
4066,DATA101,Data Science,5,S2,ETU1066,20.0
6468,GEST101,Gestion,2,S2,ETU1068,20.0
4080,DATA101,Data Science,5,S2,ETU1080,20.0
3625,ANG101,Anglais,2,S1,ETU1425,20.0
7006,GEST101,Gestion,2,S2,ETU1606,20.0


<h4 align = "center" style = "color : blue">Resume des 20 premiers etudiant ayant obtenu les faibles notes</h4>

In [94]:
TrierData.tail(20)

,course_id,matiere,credits,semestre,student_id,note
4285,DATA101,Data Science,5,S2,ETU1285,0.0
7931,PROJ101,Projet,4,S2,ETU1731,0.0
7932,PROJ101,Projet,4,S2,ETU1732,0.0
5870,ELEC101,Électricité,3,S2,ETU1270,0.0
5876,ELEC101,Électricité,3,S2,ETU1276,0.0
1627,PROG101,Programmation,5,S1,ETU1027,0.0
1629,PROG101,Programmation,5,S1,ETU1029,0.0
3376,ANG101,Anglais,2,S1,ETU1176,0.0
3377,ANG101,Anglais,2,S1,ETU1177,0.0
3390,ANG101,Anglais,2,S1,ETU1190,0.0


#### 8 - ) Calcule de  la moyenne générale de chaque étudiant.

In [95]:
Data_notes.head()

,course_id,matiere,credits,semestre,student_id,note
0,MAT101,Mathématiques,4,S1,ETU1000,6.03
1,MAT101,Mathématiques,4,S1,ETU1001,12.16
2,MAT101,Mathématiques,4,S1,ETU1002,5.80
3,MAT101,Mathématiques,4,S1,ETU1003,14.07
4,MAT101,Mathématiques,4,S1,ETU1004,7.36


La formule de calcul de moyenne General : **MoyGeneral = somme(note_matiere x Credits)/somme(Credits)**

In [96]:
SommeCredits = Data_notes["credits"].sum()
SommeCredits

np.int64(28800)

Le La somme total des credit du semestre 1 et 2 : ***Somme_Credits = 28800 credits***